### PR TITLE
Add a custom analyzers package for custom lint checks

### PR DESCRIPTION
**Pull request by josephschorr**

### CODE REVIEW

File: tools/analyzers/nilvaluecheck/nilvaluecheck.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+			inspect := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+
+			nodeFilter := []ast.Node{
+				(*ast.ReturnStmt)(nil),
+				(*ast.DeclStmt)(nil),
+			}
+
+			typePaths := strings.Split(*disalllowedPaths, ",")
+			hasTypePath := func(path string) bool {
+				for _, typePath := range typePaths {
```
Comment (jzelinskie) [inline tools/analyzers/nilvaluecheck/nilvaluecheck.go:41]: This looks like it's also stringz.SliceContains

File: tools/analyzers/nilvaluecheck/nilvaluecheck.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+						if foundType == nil {
+							return false
+						}
+
+						resultType := foundType.String()
+						if hasTypePath(resultType) {
+							// Check for a nil result.
+							returnedValue := s.Results[index]
+							if ident, ok := returnedValue.(*ast.Ident); ok {
+								if ident.Name == "nil" {
```
Comment (jzelinskie) [inline tools/analyzers/nilvaluecheck/nilvaluecheck.go:91]:
```go
if ident, ok := returnedValue.(*ast.Ident); ok && ident.Name == "nil" {
```

File: tools/analyzers/nilvaluecheck/nilvaluecheck.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+				}
+				return false
+			})
+
+			return nil, nil
+		},
+		Requires: []*analysis.Analyzer{inspect.Analyzer},
+		Flags:    *flagSet,
+	}
+}
```
Comment (jzelinskie) [inline tools/analyzers/nilvaluecheck/nilvaluecheck.go:130]: Can we add tests for this?

File: tools/analyzers/nilvaluecheck/nilvaluecheck.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+							return false
+						}
+
+						foundType := pass.TypesInfo.TypeOf(valueSpec.Type)
+						if foundType == nil {
+							return false
+						}
+
+						varType := foundType.String()
+						if len(valueSpec.Values) == 0 && hasTypePath(varType) {
```
Comment (jzelinskie) [inline tools/analyzers/nilvaluecheck/nilvaluecheck.go:112]: looks like this catches `var thing *myType`, but not `var thing *mytype = nil` or `var thing *mytype = otherNilVariable`

**Review by jzelinskie — CHANGES_REQUESTED** [review]

File: tools/analyzers/nilvaluecheck/nilvaluecheck.go (hunk 4 above)
Comment (josephschorr) [reply tools/analyzers/nilvaluecheck/nilvaluecheck.go:112]: Correct. `var thing *mytype = nil` should be caught by existing lint and to do the `otherNilVariable` would require real flow analysis

File: tools/analyzers/nilvaluecheck/nilvaluecheck.go (hunk 3 above)
Comment (josephschorr) [reply tools/analyzers/nilvaluecheck/nilvaluecheck.go:130]: Yeah, I can add some unit tests based on local parsing

**Review by jzelinskie — APPROVED** [review]

**Pull request merged**